### ISSUE TITLE
How to use Non-Default Parameter

### ISSUE THREAD

**Issue by eiswind** [Mcat12/shaku#5]
I struggle providing my ConnectionPool...
Is there a way to use something that does not implement Default as a parameter?

```
#[derive(Debug, Component)]
#[shaku(interface = UserRepo)]
pub struct UserRepository {

    pool: Arc<Pool<PostgresConnectionManager<NoTls>>>
}
```
```
 #[derive(Debug, Component)]
                      ^^^^^^^^^ the trait `std::default::Default` is not implemented for `bb8::Pool<bb8_postgres::PostgresConnectionManager<tokio_postgres::tls::NoTls>>`
```

**Comment by elertan** [comment]
Hi @eiswind .

Check this issue out, https://github.com/Mcat12/shaku/issues/1
There are 2 ways to go about this as of now.
Either you mark the default as `unreachable!` and provide the default from withing the component_with_parameters.
Or you could wrap the `Arc<...>` within an `Option` which defaults to `None`, but to me that doesn't really feel right.

Hopefully later shaku will support a better way of dealing with non defaultable parameters.

**Comment by Mcat12** [comment]
There's a third way to do this, which is to provide a valid default with `#[shaku(default = open_db_pool())]`, but that won't always be a valid workaround.

I'm interested in what kind of API you expect to make non-default parameters work. Do you expect it to be a compile-time error or runtime panic if the parameter is not passed in?

Compile-time may be quite difficult to achieve; if this is requested then I'd be interested in how you expect the API to reflect this (ex. require more parameters in the module `builder` method).

Runtime panic would be easier and is doable today. That's one of the purposes of panicking anyways, to signal a programming error. I could add an annotation like `#[shaku(no_default)]`, but I'm not sure if this makes it clear to the user that building a module with this component in it will panic if the parameter is not provided. I guess it's basically the same as the current workaround with `#[shaku(default = unreachable!())]`.

**Comment by eiswind** [comment]
I think it could happen at runtime.

The main use of DI to me is creating test setups, mocking out what I do not want to test in a specific case. In my current case I mock away the DB access to test an actix-web handler. So in this case there really is no Pool needed for the test.

My app is normally setup like this 
```
let module = MainModule::builder()
            .with_component_parameters::<DefaultPoolProvider>(DefaultPoolProviderParameters {
                pool: pool.clone()
            }).build();
```

For running the tests I mock away the PoolProvider (as it is not used anyway)
```
App::new()
                .app_data(Data::new(MainModule::builder()
                    .with_component_override::<dyn PoolProvider>(Box::new(MockPoolProvider{}))
                    .with_component_override::<dyn UserRepository>(Box::new(MockUserRepository{})).build()))
                .route("/users", web::get().to(get_all_users)),
```

 so it would not make sense to me to force require the parameter at compile time.

**Comment by Mcat12** [comment]
I've added an annotation to mark parameters without a default: `#[shaku(no_default)]`. This will be included in the next release (0.5.0), but is functionally equivalent to `#[shaku(default = unimplemented!())]`.

@eiswind hopefully we've answered your question satisfactorily? You may also be interested in using submodules to completely replace your storage layer with a mock or in-memory implementation:
https://docs.rs/shaku/0.4.1/shaku/guide/submodules/index.html

**Comment by eiswind** [comment]
@Mcat12 really thanks a lot for clearing things up! It really helped me getting my example to work. Using Submodules could help make it a little cleaner. Maybe you could add some sentence about the default to the docs, at least I didn't see it. I'll be closing this one for now.
